### PR TITLE
[go_router] adds async feature and build context to redirect

### PR DESCRIPTION
**Pull request by chunhtai**
fixes https://github.com/flutter/flutter/issues/105808
fixes https://github.com/flutter/flutter/issues/99121

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [ ] I signed the [CLA].
- [ ] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [ ] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

**Comment by chunhtai** [comment]
The ci is failing due to flutter sdk version mismatched, may need some more time to figure out, otherwise the code is ready

File: packages/go_router/README.md
Comment (loic-sharma) [inline packages/go_router/README.md:124]:
```suggestion
You can use redirection to prevent the user from visiting a specific page. In
```

File: packages/go_router/README.md
Comment (loic-sharma) [inline packages/go_router/README.md:138]:
```suggestion

If the code depends on [BuildContext](https://api.flutter.dev/flutter/widgets/BuildContext-class.html)
```

File: packages/go_router/README.md
Comment (loic-sharma) [inline packages/go_router/README.md:147]: Consider:

```suggestion
The [GoRouter.redirect](https://pub.dev/documentation/go_router/latest/go_router/GoRouter-class.html)
is always called regardless of which GoRoute was matched. The
top-level redirect always takes priority over route-level redirect.
```

File: packages/go_router/README.md
Comment (loic-sharma) [inline packages/go_router/README.md:153]:
```suggestion
GoRoute matches, e.g. GoRoute with sub-routes, the parent route redirect takes
```

File: packages/go_router/README.md
Comment (loic-sharma) [inline packages/go_router/README.md:154]: Should we reiterate that the top-level redirect will take precedence over the `GoRoute`'s redirect here? This would be repetitive, but, useful for anyone that skipped the top-level redirect section.

File: packages/go_router/example/README.md
Comment (loic-sharma) [inline packages/go_router/example/README.md:26]: I'd avoid "simple" as it can be discouraging for folks that don't find the sample simple 😄 

```suggestion
An example to demonstrate how to use redirect to handle a sign-in flow.
```

File: packages/go_router/pubspec.yaml
Comment (loic-sharma) [inline packages/go_router/pubspec.yaml:10]: Is this intended? Does this mean folks will need to use the beta/master release channel until we release the next version of Flutter?

File: packages/go_router/lib/src/router.dart
Comment (loic-sharma) [inline packages/go_router/lib/src/router.dart:34]:
```suggestion
/// during the redirection (which is how `of` methods are usually implemented),
```

File: packages/go_router/lib/src/router.dart
Comment (loic-sharma) [inline packages/go_router/lib/src/router.dart:35]: Would "reevaluation" be a better term than refresh here? Refresh makes me think the `GoRoute` will be rebuilt even if the redirect no-ops. I don't feel strongly about this though

```suggestion
/// a reevaluation will be triggered when the [InheritedWidget] changes.
```

File: packages/go_router/lib/src/redirection.dart
Comment (loic-sharma) [inline packages/go_router/lib/src/redirection.dart:15]: Looks like this comment can be deleted :)

File: packages/go_router/lib/src/redirection.dart
Comment (loic-sharma) [inline packages/go_router/lib/src/redirection.dart:30]: Can you use `await` here?

File: packages/go_router/pubspec.yaml
Comment (chunhtai) [reply packages/go_router/pubspec.yaml:10]: yes this is just a placeholder to make ci happy. go_router_v5 feature branch will merge into main after next stable release

File: packages/go_router/lib/src/redirection.dart
Comment (chunhtai) [reply packages/go_router/lib/src/redirection.dart:30]: Sychronous future can't be used in function that marked async, that is why I refactor the entire redirect pipeline.

Comment (loic-sharma) [reply packages/go_router/lib/src/redirection.dart:28]: I don't follow, this works as expected on dartpad:

```dart
import 'package:flutter/foundation.dart';

Future<int> hello() => SynchronousFuture<int>(2);

void main() async {
  print(await hello());
}
```

File: packages/go_router/example/lib/redirection.dart
Comment (loic-sharma) [inline packages/go_router/example/lib/redirection.dart:83]: Please add a comment here explaining why `SynchronousFuture` is used instead of `Future.value`

File: packages/go_router/lib/src/router.dart
Comment (loic-sharma) [inline packages/go_router/lib/src/router.dart:35]: Should this suggest using `SynchronousFuture`?

File: packages/go_router/example/lib/async_redirection.dart
Comment (loic-sharma) [inline packages/go_router/example/lib/async_redirection.dart:119]: I'd move that redirect comment up. Also, I'd capitalize all comments and add punctuation to all comments in this file for consistency.

```suggestion
                    // Log a user in and notify all listeners.
                    // This will automatically redirect from /login to / using
                    // refreshListenable.
                    StreamAuthScope.of(context).signIn('test-user');
                    setState(() {
                      loggingIn = true;
                    });
```

File: packages/go_router/example/lib/async_redirection.dart
Comment (loic-sharma) [inline packages/go_router/example/lib/async_redirection.dart:220]:
```suggestion
  /// The interval that automatically signs out the user.
  final int refreshInterval;
```

File: packages/go_router/example/lib/async_redirection.dart
Comment (chunhtai) [reply packages/go_router/example/lib/async_redirection.dart:119]: oops it was a outdated comment i forgot to delete, it no longer apply

File: packages/go_router/lib/src/redirection.dart
Comment (loic-sharma) [inline packages/go_router/lib/src/redirection.dart:20]: What do you think of making all these `FutureOr`? Benefits: 

1. Makes this breaking change a bit easier for folks migrating to v5
2. Folks don't need to do anything special for the common synchronous scenario

Drawback: our code may be a bit more complicated.

This seems like a good trade-off given folks are unlikely to use `SynchronousFuture` when they should.

Comment (chunhtai) [reply packages/go_router/lib/src/redirection.dart:20]: It is also against flutter style guide https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#avoid-using-futureort

developer can also mark a synchronous method a async if they don't want to bother with SynchronousFuture, and it is not too bad, it will be introduce a one frame delay before page transition. I don't think this is worth it to breaking flutter style guide in this case.

File: packages/go_router/lib/src/router.dart
Comment (chunhtai) [reply packages/go_router/lib/src/router.dart:34]: change to futureor

**Comment by chunhtai** [comment]
@loic-sharma PTAL :)

File: packages/go_router/README.md
Comment (loic-sharma) [inline packages/go_router/README.md:137]: Async redirects is a bit more of an advanced scenario. I'd personally prefer to put the simpler sync scenario here and keep async to an example.

File: packages/go_router/README.md
Comment (loic-sharma) [inline packages/go_router/README.md:141]:
```suggestion
(which is how `of` methods are usually implemented), the redirect will be called every time the [InheritedWidget](https://api.flutter.dev/flutter/widgets/InheritedWidget-class.html)
```

File: packages/go_router/README.md
Comment (loic-sharma) [inline packages/go_router/README.md:144]: I don't think a newcomer would understand top-level vs route-level redirects by this point. Should we update the code sample above to contain both a top-level and route-level redirect?

File: packages/go_router/example/lib/async_redirection.dart
Comment (loic-sharma) [inline packages/go_router/example/lib/async_redirection.dart:111]: This tabbing seems rather high - is that intentional?

Same comment for `App.build`

File: packages/go_router/example/lib/async_redirection.dart
Comment (loic-sharma) [inline packages/go_router/example/lib/async_redirection.dart:35]: Should we update this to use @johnpryan's `RouterConfig` changes once that's merged?

File: packages/go_router/lib/src/builder.dart
Comment (loic-sharma) [inline packages/go_router/lib/src/builder.dart:59]: Consider adding a comment as to when this scenario happens. My guess is this includes async redirect, which isn't obvious.

File: packages/go_router/test/parser_test.dart
Comment (loic-sharma) [inline packages/go_router/test/parser_test.dart:29]: This guy can also be updated to use `RouterConfig` when that's ready

**Review by loic-sharma — APPROVED** [review]
LGTM! Great work! 🚀

Please also get a review from @johnpryan as my go router knowledge has huge gaps 😅

File: packages/go_router/example/lib/async_redirection.dart
Comment (chunhtai) [reply packages/go_router/example/lib/async_redirection.dart:111]: this is using the dart format, I tried to use less tabs but it format it back.

File: packages/go_router/example/lib/async_redirection.dart
Comment (chunhtai) [reply packages/go_router/example/lib/async_redirection.dart:35]: yup

File: packages/go_router/CHANGELOG.md
Comment (johnpryan) [inline packages/go_router/CHANGELOG.md:5]:
```suggestion
  - Redesigns redirection API, adds asynchronous feature, and adds build context to redirect.
```

**Review by johnpryan — APPROVED** [review]
LGTM

**Pull request merged**
